### ISSUE TITLE
Linearizability violation with node removes/adds

### ISSUE THREAD

**Issue by aphyr** [scylladb/scylla#7243]
It looks like Scylla 4.2~rc4-0.20200915.fdf86ffb8 can exhibit linearizability violations when nodes are added or removed from the cluster, and when network partitions are involved. There might be other circumstances that lead to this bug; I'm still checking.

For example, consider [20200915T164246.000-0400.zip](https://github.com/scylladb/scylla/files/5228216/20200915T164246.000-0400.zip), where multiple keys exhibited what looks to be split-brain:

```clj
   ({:key 415, :values [[2] [16 19 20]]}
    {:key 396, :values [[1 15] [5 7 8]]}
    {:key 420, :values [[3] [5 10 11 14]]}
    {:key 418, :values [[9 16 21 30] [10 11 13 14 15]]}
    {:key 412, :values [[6 8 9] [4 5 18 19]]}
    {:key 391, :values [[3] [15 16 17]]}
    {:key 413, :values [[7 19 20 25] [8 13 14 17]]}
    {:key 411, :values [[2] [1]]})}
```

Key 391, for instance, vacillated between two different timelines. Node n4 had one version, and node n5 had a completely different one:

```clj
48	:invoke	:txn	[[:r 391 nil]]
48	:ok	:txn	[[:r 391 [3]]]
44	:invoke	:txn	[[:r 391 nil]]
44	:ok	:txn	[[:r 391 [15 16 17]]]
48	:invoke	:txn	[[:r 391 nil]]
48	:ok	:txn	[[:r 391 [3 9 10 27 28]]]
44	:invoke	:txn	[[:r 391 nil]]
44	:ok	:txn	[[:r 391 [15 16 17 24 25]]]
```

You can reproduce these issues with jepsen-scylla b3c9f5f0fd12dde0e22460f997a12d3c435d5dfa, by running
```
lein run test -w list-append --nemesis partition,remove --nemesis-interval 5 --time-limit 300 --local-deb scylla-server_4.2~rc4-0.20200915.fdf86ffb8-1_amd64.deb
```

I know there are supposed to be some extant bugs around membership changes, so feel free to flag this as a dup, or merge it into existing issues as appropriate. :)

---

This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): scylla-server_4.2~rc4-0.20200915.fdf86ffb8-1_amd64.deb
Cluster size: 5
OS (RHEL/CentOS/Ubuntu/AWS AMI): Debian Buster

**Comment by aphyr** [comment]
Looks like process pauses plus removes are sufficient as well.

**Comment by aphyr** [comment]
Ditto for process kills and removes.

**Comment by aphyr** [comment]
It's also unaffected by using write consistency `quorum`.

**Comment by aphyr** [comment]
Looks like pure decommissions will do the trick too. I think we may have moved too quick and gotten the cluster confused--it went down halfway through this test and never came back, but not before exposing nonlinearizable results. Note that we still forcibly kill/wipe nodes after issuing decommission commands, so it's possible that decommissions can be in progress when we nuke a node--this isn't a nice clean approach where we block politely on every operation. ;-)

[20200916T173447.000-0400.zip](https://github.com/scylladb/scylla/files/5235204/20200916T173447.000-0400.zip)

**Comment by kostja** [comment]
One of the problems identified by this test has been expanded in an own github issue: https://github.com/scylladb/scylla/issues/7359

**Comment by aphyr** [comment]
This issue might also reflect an invalid seed configuration used in the test: we set each node's seeds to be every node in the cluster, which is unsafe when rejoining nodes. scylla-jepsen b2d292be resolves that issue; I'm re-testing to see whether things still look wonky with the correct seed config.

**Comment by aphyr** [comment]
With correct seed configurations, I'm still seeing all kinds of anomalies in version 4.2.rc5-0.20201001.5fcc1f205c. For instance, consider this test run: 
[20201018T111331.000-0400.zip](https://github.com/scylladb/scylla/files/5397722/20201018T111331.000-0400.zip)

We get the same divegent-history behavior: elements appearing to flicker in and out of existence depending on what node you talk to:

```clj
2020-10-18 11:19:09,308{GMT}	INFO	[jepsen worker 0] jepsen.util: 120	:ok	:txn	[[:r 121 [2 3]]]
2020-10-18 11:19:11,593{GMT}	INFO	[jepsen worker 1] jepsen.util: 161	:ok	:txn	[[:r 121 [2 3 19 24 27 28]]]
2020-10-18 11:19:11,726{GMT}	INFO	[jepsen worker 4] jepsen.util: 74	:ok	:txn	[[:r 121 [19 24 27 28]]]
2020-10-18 11:19:11,853{GMT}	INFO	[jepsen worker 3] jepsen.util: 58	:ok	:txn	[[:r 121 [2 3 19 24 27 28]]]
2020-10-18 11:19:12,430{GMT}	INFO	[jepsen worker 4] jepsen.util: 74	:ok	:txn	[[:r 121 [19 24 27 28 40]]]
2020-10-18 11:19:12,853{GMT}	INFO	[jepsen worker 4] jepsen.util: 74	:ok	:txn	[[:r 121 [19 24 27 28 40]]]
2020-10-18 11:19:14,771{GMT}	INFO	[jepsen worker 4] jepsen.util: 74	:ok	:txn	[[:r 121 [19 24 27 28 40 51 52 57 61 62]]]
2020-10-18 11:19:15,783{GMT}	INFO	[jepsen worker 0] jepsen.util: 120	:ok	:txn	[[:r 121 [2 3 19 24 27 28 40 51 52 57 61 62 68 69 72 73 74]]]
```

I'm also seeing no shortage of stale reads, likely because of this split-brain weirdness.

**Comment by nyh** [comment]
Thinking about this issue led me to come up with a theoretical (not yet reproduced) bug with how list appends choose timestamps in LWT - I created issue #7611. I'm not sure if it explains this issue, but _maybe_ it's related.

**Comment by kostja** [comment]
As it happens, this issue stands for not one, but 3 distinct problems in Scylla:

* removenode is best effort https://github.com/scylladb/scylla/issues/7359
* list append operations use non-unique timeUUIDs as list keys https://github.com/scylladb/scylla/issues/7611
* topology changes are not coordinated, so it's possible to start another one without waiting for a previous one to finish (while the docs strictly prohibit this) https://github.com/scylladb/scylla/issues/7351

For the reference, all Scylla inherited from Cassandra with initial port from Java, and the upstream still has them.

The last item results from Jepsen not using a standard Scylla procedure to remove a node:

The default removenode timeout in jepsen is 5 seconds, after which the test wipes out the scylla instance: kills the process, removes scylla server directory and starts a new server in place of the removed one.
Another offence to approved Scylla operational procedures, removenode is issued while the node is still alive, so the node which is being removed can serve queries assuming an obsolete cluster topology.

At this point we have fixes for the first two items. The last one, besides having an established workaround (following the standard procedure) result in a test failure very rarely - it took me dozens of hours running the test to trigger a linearizability failure each time I tried the test with a patched Scylla. Once triggered, all studies of jepsen histories that I was able to obtain lead to #7351

To sum up, I think the issue must be closed along with #7611 and #7359. #7351 has a separate tracking id and will be closed along with Raft based toplogy changes, planned for Scylla 5.

**Comment by aphyr** [comment]
Just following up: I've tested scylla-server_4.2.1-0.20201205.eecd2ebc52-1_amd64.deb, and yeah, I can also confirm that partitions and remove-node can still drive Scylla into split-brain territory within a couple hours--only took me 1.5, rather than dozens, but that might be a HW thing--I was testing on EC2, if that helps!

Sounds like this is basically "behaving as expected": Scylla can't handle removenodes while nodes are still alive right now. I'm fine closing this (since it turned out to be multiple bugs!) and tracking further split-brain issues in #7351. :-)

```
lein run test --username admin --nodes-file /home/admin/nodes5 -w list-append --nemesis partition,remove --nemesis-interval 5 --time-limit 300 --local-deb scylla-server_4.2.1-0.20201205.eecd2ebc52-1_amd64.deb --test-count 20
```

```clj
 {:valid? false,
  :anomaly-types (:incompatible-order),
  :anomalies
  {:incompatible-order
   ({:key 29, :values [[7] [35 36 37 48 61 80 82 83 88 98]]}
    {:key 28,
     :values
     [[3 26 38 39 40] [3 23 33 51 58 59 64 65 77 85 86 99 102 103]]}
    {:key 2,
     :values
     [[3 17 19 21 22 24 32 37] [3 5 17 19 21 22 24 32 37 47 49]]})},
  :not #{:read-atomic :read-committed},
  ...
```

[20201211T222438.000Z.zip](https://github.com/scylladb/scylla/files/5681896/20201211T222438.000Z.zip)

**Comment by slivne** [comment]
Thanks @aphyr closing this as you indicated above we will keep #7351 open
